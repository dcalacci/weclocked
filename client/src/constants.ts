
enum UPLOAD_CONSTANTS {
  UPLOAD_FORM_TITLE = "Attach WeClock Export",
  UPLOAD_FORM_DESC = "Upload the .zip file from your WeClock export here. Your data will be securely shared with the WeClock team and processed into a Google Sheet visible only to you and WeClock.",
  EMAIL_TITLE = "Your Email",
  EMAIL_DESC = "A Google Sheet will be shared with this email address.",
  FORM_TITLE = "WeClock Upload",
  FORM_DESC = "Upload WeClock Export",
  FILE_SELECT = "Tap here to select a file.",
  FILE_SELECT_DRAG = "Drag and drop exports here."
}

enum STORAGE_CONSTANTS {
  EXPORTS_STORAGE_KEY = "weclock_exports",
}

export {
  UPLOAD_CONSTANTS,
  STORAGE_CONSTANTS
}
