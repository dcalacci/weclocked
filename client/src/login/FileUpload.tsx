import { createSignal, JSX } from "solid-js";
import { Show } from "solid-js";

import { UPLOAD_CONSTANTS } from "../constants";

const FileUpload = (props: {
  onFileDropped: (e: Event) => void;
  onFileChange: (e: Event) => void;
  title?: string;
  description?: string;
  uploadName?: string;
  children?: JSX.Element;
  class?: string;
}) => {
  const [dragging, setDrag] = createSignal(false);
  const isAdvancedUpload = () => {
    var div = document.createElement("div");
    return (
      ("draggable" in div || ("ondragstart" in div && "ondrop" in div)) &&
      "FormData" in window &&
      "FileReader" in window
    );
  };

  const BasicUpload = () => (
    <p
      class="transition 
    ease-in 
    shadow-md
    border-slate-400
    border-2
    cursor-pointer 
    font-semibold
    rounded-sm
    hover:shadow-sm
    hover:bg-slate-200 
    hover:text-slate-700
    active:bg-slate-300 
    active:shadow-sm
    p-1
    px-3 
    m-2"
    >
      {UPLOAD_CONSTANTS.FILE_SELECT}
    </p>
  );

  return (
    <div class={`grid grid-cols-1 space-y-2 ${props.class}`}>
      <div
        title="upload-box"
        draggable
        onDrop={(e) => {
          e.preventDefault();
          e.preventDefault();
          setDrag(false);
          props.onFileDropped(e);
        }}
        onDragEnter={() => setDrag(true)}
        onDragOver={(e) => {
          e.preventDefault();
          setDrag(true);
        }}
        onDragLeave={(e) => {
          e.preventDefault();
          setDrag(false);
        }}
        class="flex-row items-center justify-center w-full cursor-pointer"
      >
        <label
          class={`${dragging() ? "bg-gray-200" : "bg-white"} 
              transition 
              ease-in 
              flex 
              flex-col 
              rounded-sm
              border-4 
              border-dashed 
              border-slate-300
              w-full 
              h-25 
              p-5 
              group 
              text-center 
              hover:bg-slate-100
              hover:border-slate-700
              cursor-pointer`}
        >
          <div
            class="
            h-full 
              w-full 
              text-center 
              flex 
              flex-col 
              items-center 
              justify-center"
          >
            <div class="h-10 w-10">
              <svg
                xmlns="http://www.w3.org/2000/svg"
                class="w-10 h-10 text-slate-400 group-hover:text-slate-700"
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
              >
                <path
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  d="M7 16a4 4 0 01-.88-7.903A5 5 0 1115.9 6L16 6a5 5 0 011 9.9M15 13l-3-3m0 0l-3 3m3-3v12"
                />
              </svg>
            </div>
            <Show when={isAdvancedUpload()} fallback={BasicUpload}>
              <p class="text-slate-500 ">
                {UPLOAD_CONSTANTS.FILE_SELECT_DRAG}
                <br />{" "}
              </p>
            </Show>
          </div>
          <input
            title="upload-input"
            type="file"
            multiple
            class="hidden"
            onChange={props.onFileChange}
          />
          <div class="flex flex-col items-center justify-center w-full">
            {props.children}
          </div>
        </label>
      </div>
    </div>
  );
};

export default FileUpload;
