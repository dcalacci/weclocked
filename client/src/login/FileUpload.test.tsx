import { fireEvent, render, screen } from 'solid-testing-library'
import userEvent from "@testing-library/user-event";
import { UPLOAD_CONSTANTS } from "../constants";

import FileUpload from "./FileUpload"

describe('<FileUpload />', () => {
  test("renders", () => {
    const { container, unmount } = render(() => (
      <FileUpload
        onFileChange={(e) => console.log("file changed")}
        onFileDropped={(e) => console.log("file dropped")}
        title={UPLOAD_CONSTANTS.UPLOAD_FORM_TITLE}
        description={UPLOAD_CONSTANTS.UPLOAD_FORM_DESC} />
    )
    )
    expect(container.innerHTML).toMatchSnapshot()
    unmount()
  }
  )

  test("Asks to select a file", () => {
    const { container, unmount } = render(() => (
      <FileUpload
        onFileChange={(e) => console.log("file changed")}
        onFileDropped={(e) => console.log("file dropped")}
        title={UPLOAD_CONSTANTS.UPLOAD_FORM_TITLE}
        description={UPLOAD_CONSTANTS.UPLOAD_FORM_DESC} />
    )
    )
    const visibleButtonText = [UPLOAD_CONSTANTS.FILE_SELECT, UPLOAD_CONSTANTS.FILE_SELECT_DRAG]
    visibleButtonText.forEach((t) => {
      const button = screen.getByText(t)
      expect(button).toBeInTheDocument();
    })
  })

  test("User  can upload file and click upload button", async () => {
    const user = userEvent.setup()
    const { container, unmount } = render(() => (
      <FileUpload
        onFileChange={(e) => console.log("file changed")}
        onFileDropped={(e) => console.log("file dropped")}
        title={UPLOAD_CONSTANTS.UPLOAD_FORM_TITLE}
        description={UPLOAD_CONSTANTS.UPLOAD_FORM_DESC} />
    ))


    const selectFileButton = screen.getByText(UPLOAD_CONSTANTS.FILE_SELECT)
    expect(selectFileButton).toBeInTheDocument();
    await user.click(selectFileButton)

    const input = screen.getByTitle("upload-input")

    const file = new File(['hello'], 'hello.png', { type: 'image/png' })
    await userEvent.upload(input, file)
    expect(input.files[0]).toBe(file)
    expect(input.files[1]).toBe(file2)
    expect(input.files.item(0)).toBe(file)

    const file1 = new File(['hello again'], 'hello.png', { type: 'image/png' })
    const file2 = new File(['goodbye'], 'hello.png', { type: 'image/png' })
    const files = [file1, file2]
    await userEvent.upload(input, files)
    expect(input.files).toHaveLength(2)
    expect(input.files[0]).toBe(file1)
    expect(input.files[1]).toBe(file2)
    expect(input.files.item(0)).toBe(file1)
    expect(input.files.item(1)).toBe(file2)

  })
})
