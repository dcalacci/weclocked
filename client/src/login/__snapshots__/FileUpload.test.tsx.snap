// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUpload /> renders 1`] = `
"<div class=\\"grid grid-cols-1 space-y-2\\"><label class=\\"text-lg font-bold text-slate-600 tracking-wide\\">Attach WeClock Export</label><span class=\\"text-sm text-slate-500\\">Upload the .zip file from your WeClock export here. Your data will be securely shared with the WeClock team and processed into a Google Sheet visible only to you and WeClock.</span><div title=\\"upload-box\\" draggable=\\"\\" class=\\"flex items-center justify-center w-full cursor-pointer\\"><label class=\\"bg-white 
              transition 
              ease-in 
              flex 
              flex-col 
              rounded-sm
              border-4 
              border-dashed 
              border-slate-300
              w-full 
              h-60 
              p-10 
              group 
              text-center 
              hover:bg-slate-100
              hover:border-slate-700
              cursor-pointer\\"><div class=\\"h-full 
              w-full 
              text-center 
              flex 
              flex-col 
              items-center 
              justify-center\\"><div class=\\"p-10\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"w-10 h-10 text-slate-400 group-hover:text-slate-700\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\"><path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M7 16a4 4 0 01-.88-7.903A5 5 0 1115.9 6L16 6a5 5 0 011 9.9M15 13l-3-3m0 0l-3 3m3-3v12\\"></path></svg></div><div class=\\"flex flex-auto max-h-48 w-2/5 mx-auto -mt-10\\"></div><p class=\\"text-slate-500 \\">Drag and drop exports here.<br> </p><p class=\\"text-slate-500\\">or</p><p class=\\"transition 
              ease-in 
              shadow-md
              border-slate-400
              border-2
              cursor-pointer 
              font-semibold
              rounded-sm
              hover:shadow-sm
              hover:bg-slate-200 
              hover:text-slate-700
              active:bg-slate-300 
              active:shadow-sm
              p-1
              px-3 
              m-2 \\">Tap here to select a file.</p></div><input title=\\"upload-input\\" type=\\"file\\" multiple=\\"\\" class=\\"hidden\\"></label></div></div>"
`;
